compat-pack/compat-pack.md

```diff
@@ -0,0 +1,313 @@
+prem setup, making sure everything continues to work as-is. John starts by
+creating a project for the shared code, targeting .NET Standard 2.0. Given that
+the code was mostly low-level and general purpose, everything just compiled as-
+is. He then continues by creating an ASP.NET Core web application for his web
+API backend. After performing the ASP.NET Core specific changes (which was
+mostly changing namespace names) he then faces some APIs that don't exist in
+.NET Core. He runs API Port on this existing application and it points him to
+the *.NET Framework Compatibility Pack* to gain access to the missing APIs.
+His project is now compiling and he starts validating the new setup.
+
+Aspects:
+
+* Part of the application runs on .NET Framework (desktop/classic ASP.NET)
+* Other parts move to .NET Core on Windows (console/ASP.NET Core)
+* Common code is shared via .NET Standard
+
+### Building a cross-platform library
+
+Julia owns the the Fabrikam rule engine that they use for automating business
+rules. They would like to evolve it so that it can be used across all their
+applications, which include the asset management which was recently moved to
+.NET Core and now runs on a Linux machine in Azure.
+
+In order to optimize for performance, the rule engine uses both build time as
+well as runtime code generation. During build time they are using CodeDOM to
+generate entity classes out of XML. At runtime they use Reflection Emit to speed
+up evaluation of expressions in their rule engine.
+
+Neither CodeDOM nor Reflection Emit APIs are available in .NET Standard 2.0.
+During her investigation, Julia is pleased to find that these are available as
+individual packages to allow her to extend the capabilities without. This also
+allows her to check that her component only depends on APIs that work on their
+Linux-based backend as well.
+
+```XML
+<ItemGroup>
+  <PackageReference Include="System.CodeDom" Version="4.4.0" />
+  <PackageReference Include="System.Reflection.Emit" Version="4.3.0" />
+<ItemGroup>
+```
+
+Aspects:
+
+* Using the stripped down packages that only includes APIs that work on the
+  operating systems you care about.
+
+### Building .NET Standard libraries that light-up on Windows
+
+Scott has to fix a bug in Fabikam's logging infrastructure that doesn't honor
+the logging path as configured by their desktop application. For historical
+reasons, the configuration is stored in the registry. Instead of rewriting how
+configuration is persisted by the desktop application, Scott decides it's best
+to check the registry for the key, and, if present, use that to set the logging
+path. Since logging is part of their code base, Scott guards the registry call
+behind an operating system check:
+
+```CSharp
+private static string GetLoggingPath()
+{
+    // If we're on Windows and the desktop app has configured a logging path, we'll use that.
+    if (Environment.OSVersion.Platform == PlatformID.Win32NT)
```

Comment:
this should probably use `RuntimeInformation.IsOSPlatform(OSPlatform.Windows)`

Reply:
I've used `Environment.OSVersion` (1) because it's the API folks have used for years and (2) it allows for version checks. We've originally explicitly removed the `Environment.OSVersion` because version checks are known to be fragile and designed `RuntimeInformation` as the replacement. However, we've received a lot of feedback that version checks are needed in certain cases, which is why we added the API back as part of .NET Standard 2.0. At this point I see little reason to avoid the existing API.

Reply:
I understand why we added it back but that doesn't remove the reasons that it is bad to use. Going for I still think our recommendation is to use `RuntimeInformation.IsOSPlatform(OSPlatform.Windows)` and we will hopefully get back to designing better version checks into that API as well.

Reply:
What makes this particular example better?

Reply:
Will Win32NT be returned on all platforms that support registry? Maybe, maybe not. The point is that checking for an exact enum value is generally error prone as soon as we need to return a different enum value for any other reason (i.e. many Nano returns something different). That it why we designed the IsOSPlatform to support platform families or hierarchies, like unix/linux/etc.

Reply:
More generally, suppose you want to write code that does something specific on Windows and OSX: it's not possible with Environment.OSVersion.Platform today since OSX is never returned by those APIs. You'd need to mix Environment.OSVersion.Platform with RuntimeInformation which is just ugly. 

That's why we should recommend RuntimeInformation and not talk about Environment.OSVersion.Platform whenever possible :)

Reply:
Makes sense, I've changed the code and filed an issue to [deprecated the API](https://github.com/dotnet/platform-compat/issues/58).